tests/cpp/test_multidevice_tutorial.cpp

```diff
@@ -0,0 +1,367 @@
+  // Let us define a model expressing a simple memcpy
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  TensorView* tv0 = makeContigTensor(1);
+  fusion.addInput(tv0);
+  TensorView* tv1 = set(tv0); // "set" means "copy"
+  fusion.addOutput(tv1);
+
+  const bool verbose_print = verbose_ && communicator->deviceId() == 0;
+
+  if (verbose_print) {
+    // Print a concise representation of the fusion expressions
+    fusion.printMath();
+
+    // Generate and print a CUDA kernel. Notice that at this point the
+    // genereated code is just a sequential kernel as we have not
+    // scheduled the fusion yet.
+    fusion.printKernel();
+  }
+
+  // MULTIDEVICE SCHEDULING
+
+  // The first scheduling piece we need is the DeviceMesh. A device mesh in
+  // itself is nothing but a multi-dimensional array of integers representing
+  // device indices. In this tutorial, we will focus on 1D Meshes, so a
+  // DeviceMesh is nothing but a vector of indices. For example, we can consider
+  // the DeviceMesh comprised of the device index "0" only
+  DeviceMesh mesh_zero({0});
+  // or the mesh with device "1" only
+  DeviceMesh mesh_one({1});
+  // We can also consider the mesh of device indices "0" "2"
+  DeviceMesh mesh_zero_and_two({0, 2});
+  // Or the mesh containing all available devices in the world communicator
+  std::vector<int64_t> all_devices(communicator->size());
+  std::iota(
+      all_devices.begin(),
+      all_devices.end(),
+      0); // all_devices = [0,1,..., communicator->size()-1]
+  DeviceMesh mesh_full(all_devices);
+
+  // Device meshes are used to indicate the devices on which a Tensor is
+  // sharded or replicated. Each tensor can be associated with a DeviceMesh. For
+  // example,
+  tv0->setDeviceMesh(mesh_zero);
+  // means that tv0 is only present on device "0".
+  if (verbose_print) {
+    std::cout << tv0->toString() << std::endl;
+  }
+  // Alternatively:
+  tv0->setDeviceMesh(mesh_full);
+  // means that tv0 is replicated on all the devices.
+  if (verbose_print) {
+    std::cout << tv0->toString() << std::endl;
+  }
+
+  // Let us exercise this notion by examining two simple situations where we set
+  // the device mesh. In those simple examples, no network communication will be
+  // needed at all. We will simply see that only through setting the device mesh
+  // we can, e.g., replicate computation on all devices or select a subset of
+  // devices.
+
+  // FULL REPLICATION ON ALL DEVICES
+  // Let us set tv0 and tv1 to be replicated on all the devices:
+  tv0->setDeviceMesh(mesh_full);
+  tv1->setDeviceMesh(mesh_full);
+
+  // RUNTIME
+  // Set up the input
+  constexpr int64_t tensor_size = 128;
+  const c10::TensorOptions tensor_options =
+      at::TensorOptions().device(communicator->device()).dtype(at::kFloat);
+  // each rank allocate a tensor a on different device
+  at::Tensor input = at::randn({tensor_size}, tensor_options);
+  {
+    // EXECUTION
+    MultiDeviceExecutor multidevice_executor(
```

Comment:
I think it'd be helpful to mention that a single `MultiDeviceExecutor` manages a single device, i.e., a single CPU process manages a single device, which may not be obvious for users.

Reply:
ok, I added a comment